File: csharp/src/Scope.cs
namespace InariWatch.Capture;

/// <summary>
/// Per-async-context scope using AsyncLocal so each request gets its own
/// state. Mirrors Node SDK's AsyncLocalStorage and Python's contextvars.
/// </summary>
public static class Scope
{
    private static readonly AsyncLocal<ScopeData?> Current = new();

    private static readonly string[] HeaderRedactPatterns = {
        "token", "key", "secret", "auth",
        "credential", "password", "cookie", "session"
    };

    private static readonly HashSet<string> RedactBodyFields = new(StringComparer.OrdinalIgnoreCase) {
        "password", "passwd", "pass", "secret", "token",
        "api_key", "apiKey", "access_token", "accessToken",
        "refresh_token", "refreshToken", "credit_card", "creditCard",
        "card_number", "cardNumber", "cvv", "cvc", "ssn",
        "social_security", "authorization"
    };

    private static ScopeData Get()
    {
        var s = Current.Value;
        if (s != null) return s;
        s = new ScopeData();
        Current.Value = s;
        return s;
    }

    public static void SetUser(string id, string? role = null)
    {
        var u = new Dictionary<string, object> { ["id"] = id };
        if (role != null) u["role"] = role;
        Get().User = u;
    }

    public static Dictionary<string, object>? GetUser() => Current.Value?.User;

    public static void SetTag(string key, string value) => Get().Tags[key] = value;

    public static Dictionary<string, string> GetTags() =>
        new(Current.Value?.Tags ?? new Dictionary<string, string>());

    public static void SetRequestContext(Dictionary<string, object?> req) =>
        Get().Request = RedactRequest(req);

    public static Dictionary<string, object?>? GetRequestContext() => Current.Value?.Request;

    public static void Clear()
    {
        var s = Current.Value;
        if (s == null) return;
        s.User = null;
        s.Tags.Clear();
        s.Request = null;
    }

    public static void WithScope(Action body)
    {
        var prev = Current.Value;
        Current.Value = new ScopeData();
        try { body(); }
        finally { Current.Value = prev; }
    }

    public static bool ShouldRedactHeader(string name)
    {
        string lower = name.ToLowerInvariant();
        foreach (string p in HeaderRedactPatterns)
            if (lower.Contains(p)) return true;
        return false;
    }

    public static Dictionary<string, object?> RedactRequest(Dictionary<string, object?> req)
    {
        var copy = new Dictionary<string, object?>(req);
        if (copy.TryGetValue("headers", out var h) && h is Dictionary<string, object?> hMap)
        {
            var safe = new Dictionary<string, object?>(hMap.Count);
            foreach (var kv in hMap)
                safe[kv.Key] = ShouldRedactHeader(kv.Key) ? "[REDACTED]" : kv.Value;
            copy["headers"] = safe;
        }
        if (copy.TryGetValue("body", out var b) && b != null)
            copy["body"] = RedactBody(b);
        return copy;
    }

    public static object? RedactBody(object? body)
    {
        switch (body)
        {
            case null: return null;
            case string s:
                return s.Length > 1024 ? s[..1024] + "...[truncated]" : s;
            case Dictionary<string, object?> map:
                var safe = new Dictionary<string, object?>();
                foreach (var kv in map)
                {
                    safe[kv.Key] = RedactBodyFields.Contains(kv.Key)
                        ? "[REDACTED]"
                        : RedactBody(kv.Value);
                }
                return safe;
            case System.Collections.IEnumerable list when body is not string:
                var arr = new List<object?>();
                foreach (var item in list) arr.Add(RedactBody(item));
                return arr;
            default:
                return body;
        }
    }

    internal class ScopeData
    {
        public Dictionary<string, object>? User;
        public Dictionary<string, string> Tags = new();
        public Dictionary<string, object?>? Request;
    }
}

File: rust/src/scope.rs
//! Per-task / per-thread scope: user, tags, request context.
//!
//! Rust has no AsyncLocalStorage primitive, so we keep a global scope
//! protected by `parking_lot::Mutex` plus a cheap `with_scope` API that
//! takes a closure and isolates state for its duration. For tokio apps
//! `tokio::task_local!` is recommended once the `axum` feature is enabled
//! (the adapter binds the scope into the request future).

use parking_lot::RwLock;
use std::collections::HashMap;
use std::sync::Arc;

use crate::types::{RequestContext, User};

#[derive(Default)]
pub(crate) struct ScopeData {
    pub user: Option<User>,
    pub tags: HashMap<String, String>,
    pub request: Option<RequestContext>,
}

#[derive(Clone)]
pub struct Scope(pub(crate) Arc<RwLock<ScopeData>>);

impl Scope {
    pub fn new() -> Self {
        Self(Arc::new(RwLock::new(ScopeData::default())))
    }
}

impl Default for Scope {
    fn default() -> Self {
        Self::new()
    }
}

static GLOBAL_SCOPE: once_cell::sync::Lazy<Scope> = once_cell::sync::Lazy::new(Scope::new);

pub(crate) fn global_scope() -> Scope {
    GLOBAL_SCOPE.clone()
}

pub fn set_user(user: User) {
    let safe = User { id: user.id, role: user.role };
    let scope = global_scope();
    scope.0.write().user = Some(safe);
}

pub fn get_user() -> Option<User> {
    global_scope().0.read().user.clone()
}

pub fn set_tag(key: impl Into<String>, value: impl Into<String>) {
    let scope = global_scope();
    scope.0.write().tags.insert(key.into(), value.into());
}

pub fn get_tags() -> HashMap<String, String> {
    global_scope().0.read().tags.clone()
}

pub fn set_request_context(req: RequestContext) {
    let safe = redact_request(req);
    global_scope().0.write().request = Some(safe);
}

pub fn get_request_context() -> Option<RequestContext> {
    global_scope().0.read().request.clone()
}

pub fn clear_scope() {
    let scope = global_scope();
    let mut g = scope.0.write();
    g.user = None;
    g.tags.clear();
    g.request = None;
}

const HEADER_REDACT_PATTERNS: &[&str] = &[
    "token", "key", "secret", "auth",
    "credential", "password", "cookie", "session",
];

const REDACT_BODY_FIELDS: &[&str] = &[
    "password", "passwd", "pass", "secret", "token",
    "api_key", "apikey", "access_token", "accesstoken",
    "refresh_token", "refreshtoken", "credit_card", "creditcard",
    "card_number", "cardnumber", "cvv", "cvc", "ssn",
    "social_security", "authorization",
];

pub fn should_redact_header(name: &str) -> bool {
    let lower = name.to_lowercase();
    HEADER_REDACT_PATTERNS.iter().any(|p| lower.contains(p))
}

fn is_redacted_field(k: &str) -> bool {
    let lower = k.to_lowercase();
    REDACT_BODY_FIELDS.iter().any(|f| *f == lower)
}

pub fn redact_body(value: serde_json::Value) -> serde_json::Value {
    use serde_json::Value;
    match value {
        Value::String(s) => {
            if s.len() > 1024 {
                let mut out = s.chars().take(1024).collect::<String>();
                out.push_str("...[truncated]");
                Value::String(out)
            } else {
                Value::String(s)
            }
        }
        Value::Object(map) => {
            let mut safe = serde_json::Map::with_capacity(map.len());
            for (k, v) in map {
                if is_redacted_field(&k) {
                    safe.insert(k, Value::String("[REDACTED]".into()));
                } else {
                    safe.insert(k, redact_body(v));
                }
            }
            Value::Object(safe)
        }
        Value::Array(arr) => Value::Array(arr.into_iter().map(redact_body).collect()),
        other => other,
    }
}

pub fn redact_request(mut req: RequestContext) -> RequestContext {
    let scrubbed = req
        .headers
        .into_iter()
        .map(|(k, v)| {
            if should_redact_header(&k) {
                (k, "[REDACTED]".into())
            } else {
                (k, v)
            }
        })
        .collect();
    req.headers = scrubbed;

    if let Some(body) = req.body.take() {
        req.body = Some(redact_body(body));
    }
    req
}

/// Run `f` with a fresh scope, then restore the previous one. Useful for
/// short critical sections; framework adapters should swap to a
/// `task_local!`-based isolation when async work is involved.
pub fn with_scope<R>(f: impl FnOnce() -> R) -> R {
    let prev = {
        let scope = global_scope();
        let g = scope.0.read();
        ScopeData {
            user: g.user.clone(),
            tags: g.tags.clone(),
            request: g.request.clone(),
        }
    };
    clear_scope();
    let out = f();
    let scope = global_scope();
    let mut g = scope.0.write();
    *g = prev;
    out
}
